### PR TITLE
CO-112

### PR DESCRIPTION
**Pull request by JeppeTrip**

### CODE REVIEW

File: backend/Web/Services/TokenService.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    public string CreateToken(User user)
+    {
+      var key = Encoding.ASCII.GetBytes(_options.Secret);
+      var tokenHandler = new JwtSecurityTokenHandler();
+      var descriptor = new SecurityTokenDescriptor
+      {
+        Subject = new ClaimsIdentity(new Claim[]
+        {
+          new Claim(ClaimTypes.Name, user.Username)
```
Comment (CSCITM) [inline backend/Web/Services/TokenService.cs:31]: As you seem to try to lookup the claims from the token in the AuthorizationService, you should add them to the token.

File: backend/Web/Services/AuthorizationService.cs [hunk 2]
```diff
@@ -18,9 +20,13 @@ public bool IsInRole(string role)
       throw new NotImplementedException();
     }
 
-    public bool HasPolicy(string policyName)
+    public bool HasPolicy(Domain.Enums.Action policy)
     {
-      throw new NotImplementedException();
+      var policies = _httpContextAccessor.HttpContext.Request.Headers
```
Comment (CSCITM) [inline backend/Web/Services/AuthorizationService.cs:25]: You should be able to access the identity to get the list of claims if implemented correct.

File: backend/Application/Users/Commands/AssignToken/AssignTokenCommand.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      private readonly ITokenService _tokenService;
+
+      public AssignTokenCommandHandler(IApplicationDbContext context, ITokenService tokenService)
+      {
+        _context = context;
+        _tokenService = tokenService;
+      }
+      public async Task<UserTokenDto> Handle(AssignTokenCommand request, CancellationToken cancellationToken)
+      {
+        var user = await _context.Users.FirstOrDefaultAsync(x => x.Username.Equals(request.UserDto.Username) && x.Password.Equals(request.UserDto.Password));
```
[the remark this answers is not in the document]
Comment (JeppeTrip) [reply backend/Application/Users/Commands/AssignToken/AssignTokenCommand.cs:27]: For now yes. I'm aware that this is real bad, and needs to be changed to hashes and salts in the future.

File: backend/Application/Coupons/Commands/SaveCouponImage/SaveCouponImageCommand.cs [hunk 4]
```diff
@@ -9,9 +9,11 @@
 using System.Threading;
 using System.Threading.Tasks;
 using Application.Common.Exceptions;
+using Application.Common.Security;
 
 namespace Application.Coupons.Commands.SaveCouponImage
 {
+  [AuthorizeAttribute(Domain.Enums.Action.SAVE_COUPON_IMAGE)]
```
[the remark this answers is not in the document]
Comment (JeppeTrip) [reply backend/Application/Coupons/Commands/SaveCouponImage/SaveCouponImageCommand.cs:16]: It overlaps with System.Action, so I have to use it fully namespaced in places where I'm also using System.

**Comment by objmj-itminds** [comment]
I'll take it from here.

**Comment by objmj-itminds** [comment]
> 
> 
> I'll take it from here.

Or not :p

**Review by objmj-itminds — APPROVED** [review]

**Pull request merged**